### PR TITLE
[SPARK-25459][SQL] Add viewOriginalText back to CatalogTable

### PR DESCRIPTION
**Pull request by zheyuan28**
## What changes were proposed in this pull request?

The `show create table` will show a lot of generated attributes for views that created by older Spark version. This PR will basically revert https://issues.apache.org/jira/browse/SPARK-19272 back, so when you `DESC [FORMATTED|EXTENDED] view` will show the original view DDL text.

## How was this patch tested?
Unit test.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Mind filing a JIRA please?

**Comment by zheyuan28** [comment]
cc @gatorsmile

**Comment by gatorsmile** [comment]
ok to test

**Comment by gatorsmile** [comment]
This is reasonable to me. cc @jiangxb1987 @cloud-fan @hvanhovell

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 1]
```diff
@@ -467,9 +467,9 @@ private[hive] class HiveClientImpl(
         properties = filteredProperties,
         stats = readHiveStats(properties),
         comment = comment,
-        // In older versions of Spark(before 2.2.0), we expand the view original text and store
-        // that into `viewExpandedText`, and that should be used in view resolution. So we get
-        // `viewExpandedText` instead of `viewOriginalText` for viewText here.
```
Comment (jiangxb1987) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: This comment is for `viewText`, please rephrase and keep it, thanks!

**Review by jiangxb1987 — COMMENTED** [review]
LGTM only nits

**Comment by cloud-fan** [comment]
LGTM. If there are more properties like `originalViewText` which are useless to Spark and only need to be displayed, I'd suggest we create a map for them, instead of adding more and more fields into `CatalogTable`.

**Comment by dilipbiswal** [comment]
retest this please

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 2]
```diff
@@ -2348,4 +2348,17 @@ class HiveDDLSuite
...
   }
+
+  test("desc formatted table should also show viewOriginalText for views") {
+    withView("v1") {
+      sql("CREATE VIEW v1 AS SELECT 1 AS value")
+      assert(sql("DESC FORMATTED v1").collect().containsSlice(
+        Seq(
+          Row("Type", "VIEW", ""),
+          Row("View Text", "SELECT 1 AS value", ""),
+          Row("View Original Text:", "SELECT 1 AS value", "")
```
[the remark this answers is not in the document]
Comment (zheyuan28) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: @MaxGekk @gatorsmile Sorry for late, I was trying to use hive client to create the older view. But I find the [toHiveTable](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L918) method will always use the expanded view text when I create view: https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L955. Is that reasonable we change that to:
```scala
table.viewText.foreach { t => hiveTable.setViewExpandedText(t) }
table.viewOriginalText.foreach { t => hiveTable.setViewOriginalText(t) }
```
Here is my new test case:
```scala
test("SPARK-25459 desc formatted table for views created by older Spark") {
    withTable("hive_table") {
      withView("old_view") {
        spark.sql("CREATE TABLE hive_table AS SELECT 1 AS a, 2 AS b")
        val expandedView = "SELECT `gen_attr_0` AS `a`, `gen_attr_1` AS `b` FROM (SELECT " +
          "`gen_attr_0`, `gen_attr_1` FROM (SELECT `a` AS `gen_attr_0`, `b` AS " +
          "`gen_attr_1` FROM hive_table) AS gen_subquery_0) AS hive_table"
        val view = CatalogTable(
          identifier = TableIdentifier("old_view"),
          tableType = CatalogTableType.VIEW,
          storage = CatalogStorageFormat.empty,
          schema = new StructType().add("a", "int").add("b", "int"),
          viewText = Some(expandedView),
          viewOriginalText = Some("SELECT 1 AS a, 2 AS b")
        )
        hiveContext.sessionState.catalog.createTable(view, ignoreIfExists = false)
        // Check the output rows.
        assert(sql("DESC FORMATTED old_view").collect().containsSlice(
          Seq(
            Row("Type", "VIEW", ""),
            Row("View Text", expandedView, ""),
            Row("View Original Text", "SELECT 1 AS a, 2 AS b", "")
          )
        ))
      }
    }
  }
```

**Comment by dongjoon-hyun** [comment]
Hi, @zheyuan28 . Thank you for your first contribution. You've finished almost things. So, in order to be able to merge, we need the followings.
- Could you merge the Hive test case PR, https://github.com/zheyuan28/spark-1/pull/1 ?
- Could you add `[SQL]` into the title like `[SPARK-25459][SQL] ...`?

We are looking forward to seeing your more contributions in the future. Thanks.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I'll review again after your updates.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 3]
```diff
@@ -331,6 +332,7 @@ case class CatalogTable(
     comment.foreach(map.put("Comment", _))
     if (tableType == CatalogTableType.VIEW) {
       viewText.foreach(map.put("View Text", _))
+      viewOriginalText.foreach(map.put("View Original Text:", _))
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: +1 for @MaxGekk 's comment. 
@zheyuan28 .  Please review and merge, https://github.com/zheyuan28/spark-1/pull/2 .

File: sql/core/src/test/resources/sql-tests/results/higher-order-functions.sql.out [hunk 4]
```diff
@@ -201,6 +201,7 @@ struct<>
 -- !query 20 output
 
 
+
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/results/higher-order-functions.sql.out]: This one seems to be GitHub issue. When I check this PR, this file is the same with master branch.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Thank you, @zheyuan28 .
Merged to master.

**Comment by MaxGekk** [comment]
Congratulations, Chris! We are waiting for the second PR from you :-)

**Comment by HyukjinKwon** [comment]
Late LGTM too

**Pull request closed without merging**